runtime/lua/vim/treesitter/_fold.lua

```diff
@@ -0,0 +1,156 @@
+local api = vim.api
+
+local M = {}
+
+--- Memoizes a function based on the buffer tick of the provided bufnr.
+--- The cache entry is cleared when the buffer is detached to avoid memory leaks.
+---@generic F: function
+---@param fn F fn to memoize, taking the bufnr as first argument
+---@return F
+local function memoize_by_changedtick(fn)
```

Comment:
Out of scope for this PR, but this could be useful as a generic stdlib feature. Similar to python's [lru_cache](https://docs.python.org/3/library/functools.html#functools.lru_cache).